Library/OcCryptoLib/X64/Sha512Avx.nasm

```diff
@@ -0,0 +1,493 @@
+  vpxor	 xmm1, xmm1, xmm3     ; xmm1 = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8) ^ (W[t-15] SHR 7)
+
+  ; W[t] = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+  %assign idx  (%1 - 7)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7]
+  vpaddq xmm0, xmm0, xmm1     ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15])
+  %assign idx  (%1 - 16)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+  %assign idx  %1
+  vmovdqa W_t(idx), xmm0	    ; Store W[t]
+  vpaddq xmm0, xmm0, K_t(idx) ; Compute W[t]+K[t]
+  vmovdqa WK_2(idx), xmm0     ; Store W[t]+K[t] for next rounds
+%endmacro
+
+; #######################################################################
+; BOOLEAN TryEnableAvx ()
+; To run in QEMU use options: -enable-kvm -cpu Penryn,+avx,+xsave,+xsaveopt
+; #######################################################################
+align 8
+global ASM_PFX(TryEnableAvx)
+ASM_PFX(TryEnableAvx):
+  ; Detect CPUID.1:ECX.XSAVE[bit 26] = 1 (CR4.OSXSAVE can be set to 1).
+  ; Detect CPUID.1:ECX.AVX[bit 28] = 1 (AVX instructions supported).
+  mov eax, 1          ; Feature Information
+  cpuid               ; result in EAX, EBX, ECX, EDX
+  and ecx, 014000000H
+  cmp ecx, 014000000H ; check both XSAVE and AVX feature flags
+  jne noAVX
+  ; processor supports AVX instructions
+  mov rax, cr4
+  bts rax, 18  ; OSXSAVE: enables XGETBV and XSETBV
+  mov cr4, rax
+
+  mov ecx, 0          ; read the contents of XCR0 register
+  xgetbv              ; result in EDX:EAX
+  or  eax, 06H        ; enable both XMM and YMM state support
+  ; XSETBV must be executed at privilege level 0 or in real-address mode.
+  xsetbv
+  mov rax, 1
+  mov byte [rel ASM_PFX(mIsAvxEnabled)], 1
+  jmp done
+noAVX:
+  mov rax, 0
+done:
+  ret
+
+; #######################################################################
+;  void Sha512TransformAvx(sha512_state *state, const u8 *data, int blocks)
+;  Purpose: Updates the SHA512 digest stored at "state" with the message
+;  stored in "data".
+;  The size of the message pointed to by "data" must be an integer multiple
+;  of SHA512 message blocks.
+;  "blocks" is the message length in SHA512 blocks
+; #######################################################################
+align 8
+global ASM_PFX(Sha512TransformAvx)
+ASM_PFX(Sha512TransformAvx):
+  test msglen, msglen
+  je nowork
+
+  ; Allocate Stack Space
+  mov rax, rsp
+  sub rsp, frame_size
+  and rsp, ~(0x20 - 1)
+  mov [rsp + frame_RSPSAVE], rax
+
+  ; Save GPRs
+  ; Registers RBX, RBP, RDI, RSI, R12, R13, R14, R15, and XMM6-XMM15 are nonvolatile,
+  ; but XMM6-XMM15 are not used.
+  mov [rsp + frame_GPRSAVE], rbx
+  mov [rsp + frame_GPRSAVE + 8*1], rbp
+  mov [rsp + frame_GPRSAVE + 8*2], rdi
+  mov [rsp + frame_GPRSAVE + 8*3], rsi
+  mov [rsp + frame_GPRSAVE + 8*4], r12
+  mov [rsp + frame_GPRSAVE + 8*5], r13
+  mov [rsp + frame_GPRSAVE + 8*6], r14
+  mov [rsp + frame_GPRSAVE + 8*7], r15
```

Comment:
Unfortunately we cannot guarantee in UEFI world that xmm0~xmm5 actually are volatile. This is because while UEFI does follow Microsoft 64-bit ABI, it does not (officially) support vector registers as a part of the context. This means that:
- a context switch may happen and clutter the xmm registers
- another function might have forgotten about that and was itself using xmm registers when our code interrupted it.

To workaround this the following changes need to be made:
- We should save all the xmm registers we use here on the stack.
- We should restore all the saved xmm registers before exiting the function.
- We should save flags register here on the stack and disable interrupts (e.g. via cli).
- We should reenable the interrupts if they were previously enabled (can check the saved flags) in the end of the function.

Reply:
Is the vector register issue here more about security (e.g. a maliciously crafted service could interrupt this code and cause faked hashes) or more about correctness (e.g. this code gets interrupted, the registers get trashed, so the hash is incorrect)?

For the latter, is there any scope for 'working round' it by only enabling this code on a settings flag, and documenting that it _may_ be incompatible with some other drivers?

For the former, I guess we can disable some but not all interrupts? Therefore cannot completely prevent the issue? And also that the same issue would in theory apply to the pre-existing, non-avx algorithm (i.e. not a 'to spec' interrupt, but a malicious one, interrupting and changing things).

Those are just some comments on general principles - again offered somewhat tentatively - but in case anything I'm saying here seems useful to @vit9696 to feed back into what he's requesting.

Reply:
The issue is not about security: if malicious executes, we are doomed in the first place. The issue is that we need to ensure compatibility, and this sounds like the only possible way.